hummingbot/hummingbot
# Feat/bot orchestration external events

**Before submitting this PR, please make sure**:


**A description of the changes proposed in the pull request**:

This PR implements the features of Phase 2 for [HIP-Bot Orchestration](https://hummingbot-foundation.notion.site/Bot-Orchestration-fcac18bd90d74b0ebca9b260617522f0#1a435ce4e3b0410aa4a50295b91f22af).

More specifically it provides interfaces which enable consuming external events within bot instances, via a message broker. Furthermore, it allows subscribing to external topics for consuming messages.

- External Event Listener: Uses the `EEventQueueFactory` and `EEventListenerFactory` classes
- External Topic Listener: Uses the `ETopicQueueFactory` and `ETopicListenerFactory` classes

The specification defines a base URI that follows the format `{namespace}/{bot_id}/external/events/*` for consuming external events. This means that messages send to topics that follow the specification URI will be captured and consumed as external events. The name of the event is defined by the URI, for example an event arrived at `hbot/bot1/external/events/test/a` will have the name/id of `test.a`. Notice that URI slashes (`/`) are transformed to dots (`.`) when constructing the name of the event, for future multi-broker and multi-protocol support purposes.

The `EEventQueueFactory`, `EEventListenerFactory`,  `ETopicQueueFactory` and `ETopicListenerFactory` classes can be used within strategies to listen to external events and topic messages. A script example is also included (`scripts/external_events_example.py`) in this PR that demostrates usage of the provided classes.

Finally, this PR extends the global configuration and adds the `mqtt_external_events` parameter for globally enabling/disabling external events feature for bot instances.


**Tests performed by the developer**:

- [x] Tested from source installation and setup
- [x] Tested docker container installation and setup
- [x] Send  external events via MQTT and consume within a script strategy
- [x] Send  messages to costom topics via MQTT and consume within a script strategy

**Tips for QA testing**:

To test external events:

- Build this PR
- Start a bot
- Set `instance_id` parameter (e.g. `config instance_id bot1`)
- Ensure you have an MQTT broker deployed somewhere and configure parameters properly in hummingbot client
- Start mqtt bridge with `mqtt start`
- Start the `external_events_example.py` script strategy: `start --script external_events_example.py`
- Use the `external_events.py` example from the [hbot-remote-client-py](https://github.com/klpanagi/hbot-remote-client-py/blob/main/examples/external_events.py) repository to send events to the bot: `./external_events.py bot1`
